### PR TITLE
Add support for xspace

### PR DESCRIPTION
**Pull request by blipp**
The [`xspace` package](https://www.ctan.org/pkg/xspace) provides an `\xspace` command that can be used at the end of a macro. It avoids having to use `\ ` or `{}` after the macro to manually decide if a space is needed after the macro.

Example:
```
\newcommand{\foo}{SuperLongAbbreviation\xspace}
This is \foo and I like \foo.
```
which is nicer than
```
\newcommand{\foo}{SuperLongAbbreviation}
This is \foo{} and I like \foo.
```

This pull request adds support for xspace to YaLafi. It suppresses unwanted “Don't put a space before a full stop” and similar messages from LanguageTool.

I am leaving comments inline, because I am unsure at some places if the code uses the interface to YaLafi correctly.

### CODE REVIEW

File: yalafi/packages/xspace.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+xspace_excl = ['.', ',', "'", '/', '?', ';', ':', '!', '~', '-', ')', '$', '\\footnote']
+
+
+def h_xspace(parser, buf, mac, args, pos):
+    if len(args) >= 1:
+        arg = args[1]
+        if hasattr(arg[0], 'txt'):
+            if not arg[0].txt in xspace_excl:
+                arg.insert(0, defs.SpecialToken(pos+1, '\\;'))
```
Comment (blipp) [inline yalafi/packages/xspace.py:36]: `pos+1` really was only a guess. Is that a correct way to add an extra character?

File: yalafi/packages/xspace.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+require_packages = []
+
+def init_module(parser, options):
+    parms = parser.parms
+
+    macros_latex = ""
+
+    macros_python = [
+
+        Macro(parms, '\\xspace', args='OA', repl=h_xspace),
```
Comment (blipp) [inline yalafi/packages/xspace.py:18]: The LaTeX command `xspace` does not really have parameters, but here the code is claiming two (`OA`). Is this ok? It seems to work in my tests, but I am unsure if this “consumes” too much of what comes after `\xspace`.

Comment (matze-dd) [reply yalafi/packages/xspace.py:18]: Thanks a lot! Your method of inserting a new token seems to be a clever solution. I need to think about it twice.

Comment (blipp) [reply yalafi/packages/xspace.py:18]: If there is another more straightforward way to insert a new token, please let me know :) This hack with `OA` was what I came up with after some print debugging, and inspiration by the handling of `phantom`.

Comment (matze-dd) [reply yalafi/packages/xspace.py:18]: Your scheme should work well in principle. There are, however, some minor problems. For the beginning, I'd propose
```
        Macro(parms, '\\xspace', args='A', repl=h_xspace),
```
and
```
def h_xspace(parser, buf, mac, args, pos):
    arg = args[0]
    if len(arg) == 1 and type(arg[0]) is defs.VoidToken:
        return []
    if not arg[0].txt in xspace_excl:
        arg.insert(0, defs.SpecialToken(arg[0].pos, '\\;'))
    return arg
```
This does work with your tests.
- It seems that the optional argument can be omitted.
- If you request a mandatory argument, then the corresponding entry in `args` will contain at least one token. If the argument is empty (e.g., on \{\} or a paragraph break), `defs.VoidToken` is passed.
- Every token has an attribute `txt`.
- By the way, the line defining `xspace_excl` currently is longer than 80 characters.

Unfortunately, this and your current version reveal a bug in `Parser.expand_arguments()` in yalafi/parser.py. On `A{\xspace}B`, we get `A B`, since the closing \} is mistakenly assumed as argument of \\xspace. Instead, we should receive a VoidToken. I will raise a 'bug' issue.

Additionally, I was thinking about a case like
```
 \xspace {xyz\newcommand\y{Y}}
```
Firstly, the outer \{\} braces are removed during argument detection for h\_xspace(). This pushes back
```
\;xyz\newcommand\y{Y}
```
such that \\y is not restricted to the original scope anymore. I think, we can accept that. YaLafi doesn't account for scoping, anyway.
Secondly, the real \\xspace does *not* introduce space, if the next token is a '{'.
In summary, we need an information, if braces were present around the mandatory argument(s) of functions like h\_xspace(). I will raise an 'enhancement' issue for that, too.

What do you think? I would propose to first fix the bug with seeing '}' as argument, and give you a ring. After updating yalafi/parser.py, you could check that everything works. We could merge the PR and leave the problem due to a \{\}-braced argument "for further work". (The fix of the second issue will probably add an argument to the handler function interface and affect a lot of Python files. But I think, it is worth to enhance the interface with the additional information.)

**Comment by matze-dd** [comment]
I've pushed a commit that solves issue #135 (closing \} consumed as macro argument). Now, one can add tests like
```
    (r'\def\x{X\xspace}\x A', 'X A'),
    (r'\def\x{X\xspace}\x{}A', 'XA'),
    (r'\def\x{X\xspace}{\x}A', 'XA'),
```
Additionally, I'd like to add `"''"`, `'--'` and `'---'` to `xspace_excl`. This is necessary, as the scanner forms extra tokens in these cases. (I come more and more to the conclusion, that the introduction of those "special tokens" was a rather bad idea.)

If you don't mind, I could merge your PR as it currently is, and update xspace.py and test\_xspace.py afterwards in my repo.

In any case, again many thanks for this contribution to the project!

**Comment by blipp** [comment]
Thanks @matze-dd for your thorough analysis, and for finding the cases that I did not cover yet.

> If you don't mind, I could merge your PR as it currently is, and update xspace.py and test_xspace.py afterwards in my repo.

Yes, absolutely, feel free to merge and update afterwards! Thanks for being open to this pull request :)

> Additionally, I'd like to add "''", '--' and '---' to xspace_excl.

I think that makes sense. Line 243 in the xspace sources http://mirrors.ctan.org/macros/latex/required/tools/xspace.dtx defines all exceptions. Maybe these symbols are actually in this list, but I don't know – I have no idea currently what `\bgroup`, `\egroup`, etc are.

```
  ,.'/?;:!~-)\ \/\bgroup\egroup\@sptoken\space\@xobeysp
  \footnote\footnotemark
  \xspace@check@icr
```

**Comment by matze-dd** [comment]
> Thanks for being open to this pull request :)

My pleasure :)

> I think that makes sense. Line 243 in the xspace sources http://mirrors.ctan.org/macros/latex/required/tools/xspace.dtx defines 

Thanks for this reference!

> I didn't have problems with false positives from xspace, so I guess I'll continue to use it…

Personally, I don't use xspace, but it seems to be of good use for quite a lot of people!

And sorry for the noise:
It popped to my mind, that we can insert a simple `' '` space token instead of the `'\\;'` token you proposed. This runs all your tests, and we do not need to insert `'$'` in the exclusion list, which did not work correctly for `A\xspace $x$`. Perhaps, you can test your document(s) with the current version?

**Comment by blipp** [comment]
> It popped to my mind, that we can insert a simple `' '` space token instead of the `'\\;'` token you proposed. This runs all your tests, and we do not need to insert `'$'` in the exclusion list, which did not work correctly for `A\xspace $x$`. Perhaps, you can test your document(s) with the current version?

Yes, good catch. I re-ran the current master version on my large test document with plenty of xspace use, and it looks good!

**Comment by matze-dd** [comment]
> I re-ran the current master version on my large test document with plenty of xspace use, and it looks good!

Great. So, thank you again for the nice collaboration! The tool became a bit better.

For the remaining issue with `\xspace {A}`, I've opened issue #140.

**Comment by matze-dd** [comment]
> If there is another more straightforward way to insert a new token, please let me know :) 

When improving the interface to macro handler functions, I have been wondering about the argument `buf` to those functions. It was meant exactly for your application of looking ahead. During macro argument parsing, possible space already has been skipped, and `buf.cur()` returns the next token (if any) to be read from the buffer.
```
def h_xspace(parser, buf, mac, args, delim, pos):
    tok = buf.cur()
    if tok and tok.txt not in xspace_excl:
        return [defs.SpaceToken(pos, ' ')]
    return []
```
Now, `xspace_excl` has to contain `'{'` and `'}'`, too.

> Line 243 in the xspace sources http://mirrors.ctan.org/macros/latex/required/tools/xspace.dtx defines all exceptions. Maybe these symbols are actually in this list, but I don't know – I have no idea currently what \bgroup, \egroup, etc are.

These both seem to be \{ and \}, [see here](https://tex.stackexchange.com/questions/1930/when-should-one-use-begingroup-instead-of-bgroup).

I have included the above simple h\_xspace() in the current version. Could you please give me a notice, if this causes a problem?

**Comment by blipp** [comment]
> I have included the above simple h_xspace() in the current version. Could you please give me a notice, if this causes a problem?

I tested on the same document as before with the current master branch, and it worked well :)
